rest-client/src/main/java/org/jboss/pnc/restclient/AdvancedSCMRepositoryClient.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014-2020 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.restclient;
+
+import org.jboss.pnc.client.ClientException;
+import org.jboss.pnc.client.Configuration;
+import org.jboss.pnc.client.RemoteResourceException;
+import org.jboss.pnc.client.SCMRepositoryClient;
+import org.jboss.pnc.dto.notification.RepositoryCreationFailure;
+import org.jboss.pnc.dto.notification.SCMRepositoryCreationSuccess;
+import org.jboss.pnc.dto.requests.CreateAndSyncSCMRequest;
+import org.jboss.pnc.dto.response.RepositoryCreationResponse;
+import org.jboss.pnc.restclient.websocket.VertxWebSocketClient;
+import org.jboss.pnc.restclient.websocket.WebSocketClient;
+
+import java.util.concurrent.CompletableFuture;
+
+import static org.jboss.pnc.restclient.websocket.predicates.SCMRepositoryNotificationPredicates.withFailedTaskId;
+import static org.jboss.pnc.restclient.websocket.predicates.SCMRepositoryNotificationPredicates.withSuccessTaskId;
+
+public class AdvancedSCMRepositoryClient extends SCMRepositoryClient {
+
+    private WebSocketClient webSocketClient = new VertxWebSocketClient();
+
+    public AdvancedSCMRepositoryClient(Configuration configuration) {
+        super(configuration);
+        webSocketClient.connect("ws://" + configuration.getHost() + "/pnc-rest-new/notifications").join();
```

Comment:
With using the join() in the constructor, you theoretically limit reusability. If one of the wait* methods disconnects from the client, then other wait* methods won't work.(not a problem in this case) On the other hand, calling join() more then once is harmless, that's why I've been putting them inside wait* methods. 

If there is an issue with running join() concurrently, then adding `synchronized` keyword to join() method should fix that.

Reply:
Do you recommend I create 2 websocket clients then? one for each wait method?

Reply:
> @michalovjan I moved the creation of the `VertxWebSocketClient` so that it's local to the method. I found out otherwise, for most operations in SCMRepository that _does not_ involve websockets, the client would stay running and block the termination of the app because vertx is still running.
> 
> It's only when the `createAndSync` is called that the websocket client is closed, and therefore the app is terminated. So now with that tweak it's fixed. I wonder if that affects the other 'Advanced*' objects.

@thescouser89 
Yes, the circumstances are the same also for other Advanced clients... :( I guess we could create Vertx on join() if it's null (instead of the constructor of VertxWebSocketClient) and on disconnect() we close Vertx and make it null. This way Vertx is not created on WS client creation and its threads won't block CLI shutdown.

Reply:
I've made relevant comment below that could prevent this.